hack/build.sh

```diff
@@ -32,7 +32,8 @@ fi
 
 MODE="${MODE:-release}"
 GIT_COMMIT="${SOURCE_GIT_COMMIT:-$(git rev-parse --verify 'HEAD^{commit}')}"
-LDFLAGS="${LDFLAGS} -X github.com/openshift/installer/pkg/version.Raw=$(git describe --always --abbrev=40 --dirty) -X github.com/openshift/installer/pkg/version.Commit=${GIT_COMMIT}"
+GIT_TAG="${BUILD_VERSION:-${BUILD_RELEASE:-$(git describe --always --abbrev=40 --dirty)}}"
```

Comment:
In my example we have `BUILD_VERSION=v4.1.9` and `BUILD_RELEASE=201907311355` -- they work together, not as alternatives. What you'll get here is
```
GIT_TAG=v4.1.9
```
In which case this would be less confusing:
```suggestion
GIT_TAG="${BUILD_VERSION:-$(git describe --always --abbrev=40 --dirty)}"
```
But I would think you'd want `v4.1.9-201907311355`. If so, then it'd be something like:
```suggestion
if [ -z "${BUILD_VERSION}" ] || [ -z "${BUILD_RELEASE" ]; then
    GIT_TAG=$(git describe --always --abbrev=40 --dirty)
else
    GIT_TAG="${BUILD_VERSION}-${BUILD_RELEASE}"
fi
```

Reply:
> ```diff
> -GIT_TAG="${BUILD_VERSION:-${BUILD_RELEASE:-$(git describe --always --abbrev=40 --dirty)}}"
> +GIT_TAG="${BUILD_VERSION:-$(git describe --always --abbrev=40 --dirty)}"
> ```

This is effectively what we have now, and it leads to:

```console
$ wget https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.1.8/openshift-install-linux-4.1.8.tar.gz
$ tar xf openshift-install-linux-4.1.8.tar.gz
$ ./openshift-install version
./openshift-install v4.1.8-201907241243-dirty
built from commit e8d7e37ea7655522ac8f6ede471fd1d3ebd1bcba
release image quay.io/openshift-release-dev/ocp-release@sha256:3ea2648231035c1a65e8d91fa818bb225a2815bc0d6abfc35063a11eaba8659f
```

That's fine, but the `-dirty` is distracting.  And if we know that the version is v4.1.8, who cares about the timestamp?  We only ever get a single 4.1.8 far enough along to show up in front of end-users, even if there's a build hiccup or some such that causes us to build multiple rounds with the same version internally.

> ```diff
> -GIT_TAG="${BUILD_VERSION:-${BUILD_RELEASE:-$(git describe --always --abbrev=40 --dirty)}}"
> +if [ -z "${BUILD_VERSION:-}" ] \|\| [ -z "${BUILD_RELEASE:-}" ]; then
> +  GIT_TAG=$(git describe --always --abbrev=40 --dirty)
> +else
> +  GIT_TAG="${BUILD_VERSION}-${BUILD_RELEASE}"
> +fi
> ```

If we have either `BUILD_VERSION` or `BUILD_RELEASE`, I'd rather use the one we have; not fall back to the local tag if one `BUILD_*` is missing.  [The doozer source][1] has separate conditionals for each, so I don't want to assume we'll always have both available.  As I said earlier in this comment, I expect `BUILD_VERSION` to be sufficiently unique on its own.  If it's missing, the timestamp `BUILD_RELEASE` is almost certain to *also* be sufficiently unique, although it loses the 4.1 vs. 4.2 etc. distinctions.

Is that sufficient motivation for using just `BUILD_VERSION`, falling back to just `BUILD_RELEASE`, falling back to just the local tag?  Or am I missing something?

[1]: https://gitlab.cee.redhat.com/openshift-art/tools/doozer/commit/b02114a0bb26bbb7f60f1cc484c17fc3172c6df4#8f42eb7865c4c8ea16b5ede36b478344f79c992d_1240_1246

Reply:
doozer has to accommodate 3.x and 4.x which have _very_ different schemes for RELEASE. It's common for it to be unknown in 3.x (assigned during the build process). That's the reason for the conditional (would not expect it ever to be empty in 4.x).

But there will never be a situation where BUILD_VERSION is not defined in an OCP build. If just the version is good enough for you, then there's no reason to bring BUILD_VERSION into it. There'll be no difference between
```
GIT_TAG="${BUILD_VERSION:-${BUILD_RELEASE:-$(git describe --always --abbrev=40 --dirty)}}"
```
and
```
GIT_TAG="${BUILD_VERSION:-$(git describe --always --abbrev=40 --dirty)}"
```

I don't think I understand where the -dirty is coming from if not the git command. I'll run a build for ya though just for funzies...

Reply:
https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=22896476 is the build (wait for it...)

Reply:
> I don't think I understand where the -dirty is coming from if not the git command.

I think it is coming from the `git` command (because of a touched Dockerfile?), which is why I'd prefer not to be using `git` to extract the local tags when we're being called by doozer.

> If just the version is good enough for you, then there's no reason to bring `BUILD_VERSION` into it. 

Done with 59e43a3d8 -> dd7cff388.

While adjusting the commit message, I realized that using `${BUILD_VERSION}` instead of `${BUILD_VERSION}-${BUILD_RELEASE}` is going to mean a lot of nominally v4.2.0 installers for all of the 4.2 prereleases.  I'm fine with that, because we still have the source commit ID to look up the exact version used, but it makes me more open to using `${BUILD_VERSION}-${BUILD_RELEASE}` than I was earlier ;).

> https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=22896476 is the build...

Looks good:

```console
$ curl -s http://download.eng.bos.redhat.com/brewroot/work/tasks/6476/22896476/x86_64.log | grep ' GIT_TAG=v'
2019-08-02 21:12:27,911 - atomic_reactor.plugins.imagebuilder - INFO - + GIT_TAG=v4.2.0
```